Manage Servers menu

MrChip53 opened the pull request:
<!--
Ensure your title is short, descriptive, and in the imperative mood (Fix X, Change Y, instead of Fixed X, Changed Y).
For a good inspiration of what to write in commit messages and PRs please review https://chris.beams.io/posts/git-commit/ and our https://docs.jellyfin.org/general/contributing/issues.html page.
-->

**Changes**
Adds a preferences style menu to manage servers and their respective users.

Known issues:
~~The back stack.~~

To-do in different PRs:
- Return more friendly strings for last used, something like, a few seconds ago, a few minutes ago, 7 minutes ago, 2 hours ago, 10 days ago, over a month ago.
- Add user images?
- Add the add server button into preferences?

**Issues**
<!-- Tag any issues that this PR solves here.
ex. Fixes # -->

nielsvanvelzen commented:
I don't think the confirmation dialog is necessary, we've removed these kinds of dialogs from other places in the app too. When a user presses the "remove user/server" button just remove it. The "manage servers" screen could use the server address as description, additionally an "add server" button could be added. The "users screen  could use an "add user" button too, we could use the last active time as description for those users, it would also be cool if we could how profile pictures instead of the icons here. You could wrap the users in a `category {}` so they are separated from the delete server action.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/preference/dsl/OptionsLink.kt:
```diff
@@ -18,9 +20,16 @@ class OptionsLink(
+		extras = extraBundle
+	}
+
+	fun setListener(preferenceClickListener: Preference.OnPreferenceClickListener) {
```
I don't like the clicklistener in the `link` item. The link item is meant to link to other fragments. I think adding a new type called an "action" would be better.
You can copy the OptionsLink file to a new OptionsActions file and remove the fragment code from it and include the click listener.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/startup/StartupToolbarFragment.kt:
```diff
@@ -19,6 +19,11 @@ class StartupToolbarFragment : Fragment() {
 		}
 
+		binding.manageServers.setOnClickListener {
+			val activity = requireActivity()
+			if (activity is StartupActivity) activity.manageServers()
```
```suggestion
			val settingsIntent = Intent(requireActivity(), PreferencesActivity::class.java)
			settingsIntent.putExtra(PreferencesActivity.EXTRA_SCREEN, ManageServersScreen::class.qualifiedName)
			startActivity(settingsIntent)
```

The fragment shouldn't depend on a specific activity (we should fix this later for the `addServer` function too)

nielsvanvelzen commented on app/src/main/res/values/strings.xml:
```diff
@@ -433,4 +433,9 @@
     <string name="pref_authentication_link">Authentication options</string>
-</resources>
\ No newline at end of file
+    <string name="lbl_manage_servers">Manage Servers</string>
```
```suggestion
    <string name="lbl_manage_servers">Manage servers</string>
```

nielsvanvelzen commented on app/src/main/res/values/strings.xml:
```diff
@@ -433,4 +433,9 @@
-</resources>
\ No newline at end of file
+    <string name="lbl_manage_servers">Manage Servers</string>
+    <string name="lbl_manage_users">Users for %1$s</string>
```
```suggestion
    <string name="lbl_manage_users">%1$s users</string>
```

Maybe even just use the server name, not sure.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/startup/ManageServersScreen.kt:
```diff
@@ -0,0 +1,29 @@
+
+		setTitle(R.string.lbl_manage_servers)
+
+		for (entry in serverMap) {
+			link {
```
```suggestion
		serverMap.forEach(entry -> link {
```

Also more readable to call it "server" instead of "entry".

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/startup/ManageServersScreen.kt:
```diff
@@ -0,0 +1,29 @@
+	override val screen by optionsScreen {
+		val authenticationStore = get<AuthenticationStore>()
+
+		val serverMap = authenticationStore.getServers()
```
```suggestion
		val servers = authenticationStore.getServers()
```

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/startup/EditServerScreen.kt:
```diff
@@ -0,0 +1,51 @@
+		}
+
+		link {
+			setTitle(R.string.lbl_remove_server)
```
I'd use a trash icon here. Also this one could maybe use a description like "This also removes all users"

MrChip53 commented:
> I don't think the confirmation dialog is necessary, we've removed these kinds of dialogs from other places in the app too. When a user presses the "remove user/server" button just remove it. The "manage servers" screen could use the server address as description, additionally an "add server" button could be added. The "users screen could use an "add user" button too, we could use the last active time as description for those users, it would also be cool if we could how profile pictures instead of the icons here. You could wrap the users in a `category {}` so they are separated from the delete server action.

I addressed everything I believe except the user images because at this time there is no way to load anything except a resource id and the add server button which I think may be big enough to justify it own PR, either by adding an edittext pref to the Options system or making the preferences be fragments on startup actvity so we can just push the addserver frag to the top when the button is clicked

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/preference/dsl/OptionsAction.kt:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Link to a different fragment.
+ */
+class OptionsAction(
```
also kdoc is not correct now

nielsvanvelzen reviewed (changes requested):
I'm wondering if we should add another menu when clicking on users to allow a few actions:

- Sign out
- Remove

We could add some more in the future like:
- Set default
- Set as service user

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/startup/EditUserScreen.kt:
```diff
@@ -0,0 +1,66 @@
+		val account = accountManagerHelper.getAccount(userUUID)
+
+		when {
+			account?.accessToken != null -> {
```
Should be disabled instead of hidden when access token is not set

nielsvanvelzen replied:
Should use an icon/description

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/startup/EditUserScreen.kt:
```diff
@@ -0,0 +1,66 @@
+		}
+
+		action {
+			setTitle(R.string.lbl_remove)
```
Should use an icon/description

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/startup/EditUserScreen.kt:
```diff
@@ -0,0 +1,66 @@
+
+				//Pop the back stack manually
+				//Do it twice because the EditServerScreen doesn't update to remove the listed user we just deleted
+				requireActivity().supportFragmentManager.fragments[0].childFragmentManager.popBackStackImmediate()
```
How can we be certain that the first element in `fragments` is the fragment we want to pop?

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/preference/dsl/OptionsCategory.kt:
```diff
@@ -19,6 +19,16 @@ class OptionsCategory(
 		nodes.add(item)
 	}
 
+	@OptionsDSL
+	fun link(init: OptionsLink.() -> Unit) {
+		this += OptionsLink(context).apply { init() }
+	}
+
+	@OptionsDSL
+	fun action(init: OptionsAction.() -> Unit) {
+		this += OptionsAction(context).apply { init() }
+	}
+
```
These should be extension functions in their own files (OptionsLink.kt,OptionsAction.kt). Adding _all_ functions in this file will make the code too long.

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/preference/dsl/OptionsAction.kt:
```diff
@@ -0,0 +1,51 @@
+	var content: String? = null
+	var clickListener: Preference.OnPreferenceClickListener? = null
+
+	fun setListener(preferenceClickListener: Preference.OnPreferenceClickListener) {
```
Add an type alias for `Preference.OnPreferenceClickListener` so we don't leak androidx.preference classes from the DSL. Not sure if it's possible to add the typealias in the class. Extending the listener might work too if it's not final.

See https://kotlinlang.org/docs/type-aliases.html

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/preference/PreferencesActivity.kt:
```diff
@@ -20,13 +20,14 @@ class PreferencesActivity : FragmentActivity() {
 	companion object {
 		const val EXTRA_SCREEN = "screen"
 		const val EXTRA_SCREEN_ARGS = "screen_args"
+		const val MAIN_PREF_TAG = "PreferencesActivity"
```
```suggestion
		const val FRAGMENT_TAG = "PreferencesActivity"
```

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/preference/dsl/OptionsLink.kt:
```diff
@@ -18,9 +20,11 @@ class OptionsLink(
 	var icon: Int? = null
 	var content: String? = null
 	var fragment: KClass<out OptionsFragment>? = null
+	var extras: Bundle = bundleOf()
```
```suggestion
	var extras: Bundle? = null
```

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/preference/dsl/OptionsLink.kt:
```diff
@@ -18,9 +20,11 @@ class OptionsLink(
+	var extras: Bundle = bundleOf()
 
-	inline fun <reified T : OptionsFragment> withFragment() {
+	inline fun <reified T : OptionsFragment> withFragment(extraBundle: Bundle = bundleOf()) {
```
```suggestion
	inline fun <reified T : OptionsFragment> withFragment(extraBundle: Bundle? = null) {
```

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/preference/dsl/OptionsLink.kt:
```diff
@@ -36,6 +40,7 @@ class OptionsLink(
 			it.isPersistent = false
 			it.key = UUID.randomUUID().toString()
 			it.fragment = fragment?.qualifiedName
+			it.extras.putAll(extras)
```
```suggestion
			extras?.let { extras -> it.extras.putAll(extras) }
```

nielsvanvelzen commented on app/src/main/java/org/jellyfin/androidtv/ui/preference/dsl/OptionsLink.kt:
```diff
@@ -49,8 +54,3 @@ class OptionsLink(
 		}
 	}
 }
-
-@OptionsDSL
-fun OptionsCategory.link(init: OptionsLink.() -> Unit) {
-	this += OptionsLink(context).apply { init() }
-}
```
Put this one back (see other comment in OptionsCategory.kt)

nielsvanvelzen commented on app/src/main/res/values/strings.xml:
```diff
@@ -433,4 +433,16 @@
\ No newline at end of file
+    <string name="lbl_manage_servers">Manage servers</string>
+    <string name="lbl_remove_users">This will also remove all associated users</string>
+    <string name="lbl_remove_user_content">Remove this account</string>
```
```suggestion
    <string name="lbl_remove_user_content">Remove account</string>
```

nielsvanvelzen commented on app/src/main/res/values/strings.xml:
```diff
@@ -433,4 +433,16 @@
+    <string name="lbl_manage_servers">Manage servers</string>
+    <string name="lbl_remove_users">This will also remove all associated users</string>
+    <string name="lbl_remove_user_content">Remove this account</string>
+    <string name="lbl_sign_out_content">Sign this user out</string>
```
```suggestion
    <string name="lbl_sign_out_content">Sign this account out.</string>
```

MrChip53 reviewed (commented):
Your changes LGTM @nielsvanvelzen

thornbill reviewed (changes requested):
I tested this a bit and it looks great!

I did notice an issue though where I was unable to remove a server. I'm guessing this was the server that I last logged into. It seems to keep a reference to the last logged in user even after removing the server, so it doesn't actually go away.

nielsvanvelzen commented:
I think that behavior _should_ be fixed with #730

thornbill commented:
Whoops there's another conflict here now.
